proc.c

```diff
@@ -1456,7 +1485,14 @@ rb_hash_proc(st_index_t hash, VALUE prc)
         rb_bug("rb_hash_proc: unknown block type %d", vm_block_type(&proc->block));
     }
 
-    return rb_hash_uint(hash, (st_index_t)proc->block.as.captured.ep);
+    /* ifunc procs have their own allocated ep. If an ifunc is duplicated, they
+     * will point to different ep but they should return the same hash code, so
+     * we cannot include the ep in the hash. */
+    if (vm_block_type(&proc->block) != block_type_ifunc) {
+        hash = rb_hash_uint(hash, (st_index_t)proc->block.as.captured.ep);
+    }
```

Comment:
This seems like it'd introduce a hash collision for all the ifunc procs that have the same func and data. Maybe we need to hash the contents of ep (VM_ENV_DATA_SIZE)?

Reply:
I don't think there's much benefits to it. All `ifunc` have the same values for the four ep attributes except for ones created from `Method#to_proc`, and we have to make sure we don't use `VM_ENV_DATA_INDEX_ENV` since it's an object that can change in value during compaction.